src/app/addons/paymentwall/controllers/common/paymentwall.php

```diff
@@ -0,0 +1,141 @@
+<?php
+if (!defined('BOOTSTRAP')) {
+    die('Access denied');
+}
+
+use Tygh\Registry;
+
+require_once Registry::get('config.dir.lib') . DS . 'other' . DS . 'paymentwall-php' . DS . 'lib' . DS . 'paymentwall.php';
+
+if ($mode == 'ajax') {
+    $status = db_get_field("SELECT status FROM ?:orders WHERE order_id = ?i", $_REQUEST['order_id']);
+    if ($status == "P") {
+        fn_clear_cart($_SESSION['cart']);
+    }
+    die($status);
+}
+
+if ($mode == 'frame') {
+    $view->assign('orderId', $_REQUEST['order_id']);
+}
+
+// Payment widget
+if ($mode == 'payment') {
+
+    $rid = $_REQUEST['order_id'];
+    $sid = $_SESSION['pw_order_id'];
+    $iframe = '';
+    $baseUrl = fn_url();
+    $matchOrder = ($rid == $sid);
+    $orderId = $sid;
+    $orderInfo = fn_get_order_info($orderId);
+
+    if ($matchOrder && $orderInfo) {
+
+        $coefficient = db_get_field("SELECT coefficient FROM ?:currencies WHERE currency_code = ?s", $orderInfo['secondary_currency']);
+        $realPrice = $orderInfo['total'] / $coefficient;
+        $paymentInfo = getPaymentConfigs($orderInfo['payment_id']);
+
+        // Prepare Widget Data
+        $uid = empty($orderInfo['user_id']) ? $orderInfo['ip_address'] : $orderInfo['user_id'];
+        $widgetCode = $paymentInfo['widget_type'];
+        $currencyCode = $orderInfo['secondary_currency'];
+        $testMode = $paymentInfo['test_mode'];
+        $products = array();
+
+        Paymentwall_Config::getInstance()->set(array(
+            'api_type' => Paymentwall_Config::API_GOODS,
+            'public_key' => $paymentInfo['key'],
+            'private_key' => $paymentInfo['secret']
+        ));
+
+        if (count($orderInfo['products']) > 0) {
+            $products[] = new Paymentwall_Product($orderId, $realPrice, $currencyCode, 'Order #' . $orderId);
+        }
+
+        $widget = new Paymentwall_Widget($uid, $widgetCode, $products, array(
+            'email' => $orderInfo['email'],
+            'payment_id' => $orderInfo['payment_id'],
+            'integration_module' => 'cs_cart',
+            'test_mode' => $testMode,
+            'ref' => rand(99, 999)

```

Comment:
U sure?
I see some code in pw lib: class Paymentwall_Pingback and isSignatureValid method
```php
} else if ($this->getApiType() == Paymentwall_Config::API_GOODS) {
    $signatureParams = array('uid', 'goodsid', 'slength', 'speriod', 'type', 'ref');
}
```